From b161bf6a6c6841000720e05d527dd3a2a83802b6 Mon Sep 17 00:00:00 2001
From: Corentin Wallez <cwallez@chromium.org>
Date: Thu, 6 Jun 2024 08:06:00 +0000
Subject: //gpu: Update casts for Dawn's getProc injection before removal.

The type aliases used as the type for getProc in
dawn::native::opengl::RequestAdapterOptionsGetGLProc is being updated to
better reflect the type in EGL. Changing it breaks due to the existing
casts in Chromium casting to a void(*)(const char*) explicitly. Use
decltype(getProc) instead for now, while Dawn is updated.

Once Dawn is updated and rolled, the casts will be removed.

Bug: 343870490
Change-Id: I4d0300390a6259eefbd19ad40d6145a14c0f7ada
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5600765
Reviewed-by: Colin Blundell <blundell@chromium.org>
Commit-Queue: Colin Blundell <blundell@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1311115}
---
 .../ahardwarebuffer_image_backing_factory_unittest.cc     | 4 +++-
 gpu/command_buffer/service/webgpu_decoder_impl.cc         | 4 +++-
 gpu/config/gpu_info_collector.cc                          | 8 ++++++--
 3 files changed, 12 insertions(+), 4 deletions(-)

diff --git a/gpu/command_buffer/service/shared_image/ahardwarebuffer_image_backing_factory_unittest.cc b/gpu/command_buffer/service/shared_image/ahardwarebuffer_image_backing_factory_unittest.cc
index 2e9a381a77649..8fff071f02c93 100644
--- a/gpu/command_buffer/service/shared_image/ahardwarebuffer_image_backing_factory_unittest.cc
+++ b/gpu/command_buffer/service/shared_image/ahardwarebuffer_image_backing_factory_unittest.cc
@@ -211,8 +211,10 @@ TEST_F(AHardwareBufferImageBackingFactoryTest, ProduceDawnOpenGLES) {
 
   dawn::native::opengl::RequestAdapterOptionsGetGLProc
       adapter_options_get_gl_proc = {};
+  // TODO(343870490): Remove the cast once the type is updated in Dawn.
   adapter_options_get_gl_proc.getProc =
-      reinterpret_cast<void* (*)(const char*)>(gl::GetGLProcAddress);
+      reinterpret_cast<decltype(adapter_options_get_gl_proc.getProc)>(
+          gl::GetGLProcAddress);
   gl::GLDisplayEGL* gl_display = gl::GLSurfaceEGL::GetGLDisplayEGL();
   if (gl_display) {
     adapter_options_get_gl_proc.display = gl_display->GetDisplay();
diff --git a/gpu/command_buffer/service/webgpu_decoder_impl.cc b/gpu/command_buffer/service/webgpu_decoder_impl.cc
index 975dc227956d4..df9c2f2c7f363 100644
--- a/gpu/command_buffer/service/webgpu_decoder_impl.cc
+++ b/gpu/command_buffer/service/webgpu_decoder_impl.cc
@@ -1701,8 +1701,10 @@ wgpu::Adapter WebGPUDecoderImpl::CreatePreferredAdapter(
 #if BUILDFLAG(USE_DAWN) && BUILDFLAG(DAWN_ENABLE_BACKEND_OPENGLES)
   dawn::native::opengl::RequestAdapterOptionsGetGLProc
       adapter_options_get_gl_proc = {};
+  // TODO(343870490): Remove the cast once the type is updated in Dawn.
   adapter_options_get_gl_proc.getProc =
-      reinterpret_cast<void* (*)(const char*)>(gl::GetGLProcAddress);
+      reinterpret_cast<decltype(adapter_options_get_gl_proc.getProc)>(
+          gl::GetGLProcAddress);
   gl::GLDisplayEGL* gl_display = gl::GLSurfaceEGL::GetGLDisplayEGL();
   if (gl_display) {
     adapter_options_get_gl_proc.display = gl_display->GetDisplay();
diff --git a/gpu/config/gpu_info_collector.cc b/gpu/config/gpu_info_collector.cc
index f23793ac1c5c9..2ad07cdb31162 100644
--- a/gpu/config/gpu_info_collector.cc
+++ b/gpu/config/gpu_info_collector.cc
@@ -426,8 +426,10 @@ void ReportWebGPUSupportMetrics(dawn::native::Instance* instance) {
 
   dawn::native::opengl::RequestAdapterOptionsGetGLProc
       adapter_options_get_gl_proc = {};
+  // TODO(343870490): Remove the cast once the type is updated in Dawn.
   adapter_options_get_gl_proc.getProc =
-      reinterpret_cast<void* (*)(const char*)>(gl::GetGLProcAddress);
+      reinterpret_cast<decltype(adapter_options_get_gl_proc.getProc)>(
+          gl::GetGLProcAddress);
   adapter_options_get_gl_proc.display = EGL_NO_DISPLAY;
   adapter_options_get_gl_proc.nextInChain = adapter_options.nextInChain;
   adapter_options.nextInChain = &adapter_options_get_gl_proc;
@@ -916,8 +918,10 @@ void CollectDawnInfo(const gpu::GpuPreferences& gpu_preferences,
 #if BUILDFLAG(DAWN_ENABLE_BACKEND_OPENGLES)
   dawn::native::opengl::RequestAdapterOptionsGetGLProc
       adapter_options_get_gl_proc = {};
+  // TODO(343870490): Remove the cast once the type is updated in Dawn.
   adapter_options_get_gl_proc.getProc =
-      reinterpret_cast<void* (*)(const char*)>(gl::GetGLProcAddress);
+      reinterpret_cast<decltype(adapter_options_get_gl_proc.getProc)>(
+          gl::GetGLProcAddress);
   adapter_options_get_gl_proc.display = EGL_NO_DISPLAY;
   adapter_options_get_gl_proc.nextInChain = adapter_options.nextInChain;
   adapter_options.nextInChain = &adapter_options_get_gl_proc;
-- 
2.43.0

